kolibri/content/utils/channel_import.py

```diff
@@ -0,0 +1,287 @@
+                    return ids[middle] + 1
+                return find_hole_in_list(ids[middle:])
+            # We should only reach this point in the first iteration, if there are no holes in either
+            # the first or the last half of the list, therefore, we just take the max of the list plus 1
+            # Because the list is already sorted, we can just take the last value
+            return ids[-1] + 1
+        return find_hole_in_list(tree_ids)
+
+    def generate_row_mapper(self, mappings=None):
+        if mappings is None:
+            mappings = {}
+
+        def mapper(record, column):
+            if column in mappings:
+                col_map = mappings.get(column)
+                if hasattr(record, col_map):
+                    return getattr(record, col_map)
+                elif hasattr(self, col_map):
+                    return getattr(self, col_map)(record)
+                else:
+                    raise AttributeError('Column mapping specified but no valid column name or method found')
+            else:
+                return getattr(record, column, None)
+        return mapper
+
+    def base_table_mapper(self, SourceRecord):
+        for record in self.source.session.query(SourceRecord).all():
+            yield record
+
+    def generate_table_mapper(self, mapping=None):
+        if mapping is None:
+            return self.base_table_mapper
+        # Can only be a method on the Import object
+        if hasattr(self, mapping):
+            return getattr(self, mapping)
+        raise AttributeError('Table mapping specified but no valid method found')
+
+    def table_import(self, model, row_mapper, table_mapper, unflushed_rows):
+        DestinationRecord = self.destination.get_class(model)
+        dest_table = DestinationRecord.__table__
+
+        # If the source class does not exist (i.e. this table is undefined in the source database)
+        # this will raise an error so we set it to None. In this case, a custom table mapper must
+        # have been set up to handle the fact that this is None.
+        try:
+            SourceRecord = self.source.get_class(model)
+        except ClassNotFoundError:
+            SourceRecord = None
+
+        columns = dest_table.columns.keys()
+        data_to_insert = []
+        merge = model in merge_models
+        for record in table_mapper(SourceRecord):
+            data = {
+                str(column): row_mapper(record, column) for column in columns if row_mapper(record, column) is not None
+            }
+            if merge:
+                # Models that should be merged (see list above) need to be individually merged into the session
+                # as SQL Alchemy ORM does not support INSERT ... ON DUPLICATE KEY UPDATE style queries,
+                # as not available in SQLite, only MySQL as far as I can tell:
+                # http://hackthology.com/how-to-compile-mysqls-on-duplicate-key-update-in-sql-alchemy.html
```

Comment:
Yeah, but there's no standard SQL way to do it, so unless we add dialect specific code for each dialect we could use, this is the safest way to encode it.